### ISSUE TITLE
Improve S4158 - Empty Collections should not be Enumerated

### ISSUE THREAD

**Issue by Tim-Pohlmann** [SonarSource/sonar-dotnet#7457]
Continuation to #7300 

* [x] Unlearn NotEmpty from `list.Remove(x)` etc.
* [ ] Learn Empty from
  * [x] `list.Clear()`
  * [x] `list.Count == 0`, `list.Count() == 0` etc.
  * [ ] `list.RemoveAll(x => true)`, `set.RemoveWhere(x => true)` and `set.IntersectWith(empty)`
* [x] Learn NotEmpty from `list.Any()`, `list.Exists(5)`, `list.Contains(5)` etc.
* [ ] Support well-known extension and static methods (e.g. LINQ, `Array.BinarySearch(empty, 5)`...) - here is only needed to figure which are these methods, add their names to the collections of methods that we track and add UTs.
* [ ] Raise on `.Count()`, `.Count` and `.Length`
* [ ] Support more collections (e.g. non-generic collections, StringBuilder, Span)
* [ ] Support `FlowCaptureReference` like `(condition ? list : list).Add(1)`
* [ ] Pass constraints via `.ToArray()`, `.AsReadOnly()` etc.
* [ ] Support `Array.Empty` and `Enumerable.Empty`
* [ ] Look for `NumberConstraint` when deciding whether `new int[value]` is empty or not
* [ ] Unlearn all collections constraints from `HashSet<int>.SymmetricExceptWith` (are there similar methods?)
* [ ] Learn `CollectionConstraint` [less agressively ](https://github.com/SonarSource/sonar-dotnet/blob/ecf5aea33b61117f7024561cfc9e4910c64b83f4/analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CollectionTracker.cs#L158)for `.AddRange` and [other methods](https://github.com/SonarSource/sonar-dotnet/blob/ecf5aea33b61117f7024561cfc9e4910c64b83f4/analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CollectionTracker.cs#L47) that may or may not add elements
* [ ] Add more tests:
  * Re-assigning the collection to a new variable
  * `list?.Clear()`
  * `(list as IList<int>).Clear()`
  * loops that only add items to a list on later iterations 
  * foreach loop (call to GetEnumerator)

**Comment by Tim-Pohlmann** [comment]
Some of these tasks will be easier and more impactful after #7866

**Comment by CristianAmbrosini** [comment]
We should also add C#13 LINQ methods: CountBy, AggregateBy, and Index (see the documentation [here](https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-9/overview#net-libraries)).

```cs
List<int> emptyList = [];
_ = emptyList.CountBy(x => x);  // FN
List<int> emptyList2 = [];
_ = emptyList2.AggregateBy(x => x, x => x, (x, y) => x + y); // FN
List<int> emptyList3 = [];
_ = emptyList3.Index();         // FN
```

**Comment by martin-strecker-sonarsource** [comment]
I don't know if this is already covered or not, but collection expressions should also be supported:
```cs
object[] os1 = [];   // Learn empty
object[] os2 = [1]; // Learn not empty
object[] os3 = [..os1]; // Learn empty
// etc.
```

**Comment by Tim-Pohlmann** [comment]
@martin-strecker-sonarsource https://github.com/SonarSource/sonar-dotnet/issues/8539

**Comment by martin-strecker-sonarsource** [comment]
Tracked by https://sonarsource.atlassian.net/browse/NET-1188 now
